### PR TITLE
[examples] Added notifications with AST contents on document change

### PR DESCRIPTION
**Pull request by spoenemann**
This can be used to visualize AST contents.

### CODE REVIEW

File: packages/langium/src/workspace/document-builder.ts [hunk 1]
```diff
@@ -59,7 +59,7 @@ export interface DocumentBuilder {
 }
 
 export type DocumentUpdateListener = (changed: URI[], deleted: URI[]) => void
-export type DocumentBuildListener = (built: LangiumDocument[], cancelToken: CancellationToken) => Promise<void>
+export type DocumentBuildListener = (built: LangiumDocument[], cancelToken: CancellationToken) => void | Promise<void>
```
Comment (Lotes) [inline packages/langium/src/workspace/document-builder.ts:62]: Is `void | ` really needed? It is used as callback, which you normally never await.

Comment (Lotes) [reply packages/langium/src/workspace/document-builder.ts:62]: Or differently: Are you now not forced to check at every usage if it returns a Promise or not?

Comment (spoenemann) [reply packages/langium/src/workspace/document-builder.ts:62]: No, you can just `await` the result regardless whether it returns a `Promise` or `undefined`. If you call `await` with something that is not `Thenable`, it just returns that value immediately.

File: examples/domainmodel/src/language-server/main-browser.ts [hunk 2]
```diff
@@ -15,7 +15,21 @@ const messageWriter = new BrowserMessageWriter(self);
...
 
 // Inject the shared services and language-specific services
-const { shared } = createDomainModelServices({ connection, ...EmptyFileSystem });
+const { shared, domainmodel } = createDomainModelServices({ connection, ...EmptyFileSystem });
 
 // Start the language server with the shared services
 startLanguageServer(shared);
+
+// Send a notification with the serialized AST after every document change
+type DocumentChange = { uri: string, content: string };
```
[the remark this answers is not in the document]
Comment (spoenemann) [reply examples/domainmodel/src/language-server/main-browser.ts:24]: But where to put such centralized code? I don't want to pollute the core package with example-related code, and I'm too lazy to maintain an additional package for it. As long as it's not too much code, I'm fine with a bit of copy & paste.

See also #239.

File: packages/langium/src/workspace/document-builder.ts (hunk 1 above)
Comment (Lotes) [reply packages/langium/src/workspace/document-builder.ts:62]: Oh ok, thanks for filling this knowledge gap :D

File: examples/domainmodel/src/language-server/main-browser.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Lotes) [reply examples/domainmodel/src/language-server/main-browser.ts:24]: Yes, true. A own package would make sense - in the future. Like having scoping strategy A, B, C or extra services, that makes the Langium experience like playing with Lego bricks.

[the remark this answers is not in the document]
Comment (spoenemann) [reply examples/domainmodel/src/language-server/main-browser.ts:24]: Yes, let's keep that in mind and start such a package once we reach a "critical mass".

File: examples/statemachine/src/language-server/main-browser.ts [hunk 3]
```diff
@@ -15,7 +15,21 @@ const messageWriter = new BrowserMessageWriter(self);
...
 startLanguageServer(shared);
+
+// Send a notification with the serialized AST after every document change
+type DocumentChange = { uri: string, content: string };
+const documentChangeNotification = new NotificationType<DocumentChange>('browser/DocumentChange');
+const jsonSerializer = statemachine.serializer.JsonSerializer;
+shared.workspace.DocumentBuilder.onBuildPhase(DocumentState.Validated, documents => {
+    for (const document of documents) {
+        const json = jsonSerializer.serialize(document.parseResult.value);
+        connection.sendNotification(documentChangeNotification, {
```
Comment (Lotes) [inline examples/statemachine/src/language-server/main-browser.ts:30]: What about errors in the AST? Do you also want the user aware of validation, syntax and lexer errors?

Comment (spoenemann) [reply examples/statemachine/src/language-server/main-browser.ts:30]: Those are already sent here:
https://github.com/langium/langium/blob/e24dee8a4085e520c8ac319e836caa6c596fb6ac/packages/langium/src/lsp/language-server.ts#L206-L221

Comment (Lotes) [reply examples/statemachine/src/language-server/main-browser.ts:30]: Oki doki!

**Review by kaisalmen — APPROVED** [review]
LGTM

**Pull request merged**